src/include/migraphx/heap.hpp

```diff
@@ -0,0 +1,49 @@
+#ifndef MIGRAPHX_GUARD_RTGLIB_HEAP_HPP
+#define MIGRAPHX_GUARD_RTGLIB_HEAP_HPP
+
+#include <thread>
+#include <cmath>
+#include <algorithm>
+#include <vector>
+#include <cassert>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template <class T, class Compare>
+struct heap
+{
+    std::vector<T> data;
+    Compare compare;
+
+    heap(const std::vector<T>& val, Compare comp) : data(val), compare(std::move(comp))
+    {
+        for(int i = 2; i <= data.size(); ++i)
+        {
+            std::push_heap(data.begin(), data.begin() + i, compare);
+        }
+    }
+
+    void update(const T val)
+    {
+        std::pop_heap(data.begin(), data.end(), compare);
+        if(compare(val, data.back()))
```

Comment:
Shouldnt we do `if(compare(val. data.front())` before calling `pop_heap`?

Reply:
No need to do that, the if(compare(val, data.back())) is the same as what you mention before pop_back().

Reply:
Yes, but it requires swapping the values, and then pushing it back on the heap. We can avoid popping and pushing to the heap if we know the value wont be added.

Reply:
move the comparison to the start of the function.